libshaderc_spvc/src/spvc.cc

```diff
@@ -0,0 +1,148 @@
+// Copyright 2018 The Shaderc Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "shaderc/spvc.h"
+#include "libshaderc_util/exceptions.h"
+
+#include "spirv-tools/libspirv.hpp"
+#include "spirv_glsl.hpp"
+#include "spirv_hlsl.hpp"
+#include "spirv_msl.hpp"
+
+struct shaderc_spvc_compiler {};
+
+// Described in spvc.h.
+struct shaderc_spvc_compilation_result {
+  std::string validation_messages;
+  std::string compiler_output;
+  shaderc_compilation_status status =
+      shaderc_compilation_status_null_result_object;
+};
+
+struct shaderc_spvc_compile_options {
+  bool validate = true;
+  spv_target_env target_env = SPV_ENV_VULKAN_1_0;
+  spirv_cross::CompilerGLSL::Options glsl;
+  spirv_cross::CompilerHLSL::Options hlsl;
+  spirv_cross::CompilerMSL::Options msl;
+};
+
+shaderc_spvc_compile_options_t shaderc_spvc_compile_options_initialize() {
+  return new (std::nothrow) shaderc_spvc_compile_options;
+}
+
+shaderc_spvc_compile_options_t shaderc_spvc_compile_options_clone(
+    shaderc_spvc_compile_options_t options) {
+  if (options) return new (std::nothrow) shaderc_spvc_compile_options(*options);
+  return shaderc_spvc_compile_options_initialize();
+}
+
+void shaderc_spvc_compile_options_release(
+    shaderc_spvc_compile_options_t options) {
+  delete options;
+}
+
+void shaderc_spvc_compile_options_set_target_env(
+    shaderc_spvc_compile_options_t options, shaderc_target_env target,
+    uint32_t version) {
```

Comment:
Would it be better to use the enum type `shaderc_env_version` instead of a uint32_t for the version?  The same would go for other places you use the version.

Reply:
The other 'version' is for language version, and it looks like shaderc, glslang, spirv-cross all use a number for that rather than an enum.  I used uint32_t to match what spirv-cross uses.

Reply:
Yep.  It's traditional in this space to use a 32-bit int.  Even though that's begging for misinterpretation and name clashes.  At this point I go with the flow.